File: 811. Subdomain Visit Count.py
from collections import defaultdict
class Solution:
    def subdomainVisits(self, cpdomains: List[str]) -> List[str]:
        dic = defaultdict(int)
        for item in cpdomains:
            (count, web) = item.split()
            weblist = web.split('.')
            dom = ''
            for i in weblist[::-1]:
                dom = i + '.' + dom
                dic[dom] += int(count)
        return [' '.join([str(dic[key]), key])[:-1] for key in dic]

File: 811. Subdomain Visit Count.java
class Solution {
    public List<String> subdomainVisits(String[] cpdomains) {
        Map<String, Integer> map = new HashMap<>();
        for (String s : cpdomains) {
            String[] ss = s.split(" ");
            int count = Integer.parseInt(ss[0]);
            String domain = ss[1];
            map.put(domain, map.getOrDefault(domain, 0)+count);
            int index;
            while ((index = domain.indexOf('.')) != -1) {
                domain = domain.substring(index+1);
                map.put(domain, map.getOrDefault(domain, 0)+count);
            }
        }
        List<String> ans = new ArrayList<>();
        for (String s : map.keySet()) {
            ans.add(map.get(s) + " " + s);
        }
        return ans;
    }
}